pkg/virt-operator/resource/apply/apiservices.go

```diff
@@ -15,68 +16,77 @@ import (
+	var cachedAPIService *apiregv1.APIService
+	var err error
+	obj, exists, _ := r.stores.APIServiceCache.Get(apiService)
+	// since these objects was in the past unmanaged, reconcile and pick it up if it exists
+	if !exists {
+		cachedAPIService, err = r.aggregatorclient.Get(context.Background(), apiService.Name, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			exists = false
+		} else if err != nil {
+			return err
+		} else {
+			exists = true
 		}
+	} else if exists {
+		cachedAPIService = obj.(*apiregv1.APIService)
+	}
 
-		certsMatch := true
-		if exists {
-			if !reflect.DeepEqual(apiService.Spec.CABundle, cachedAPIService.Spec.CABundle) {
-				certsMatch = false
-			}
+	if !exists {
+		r.expectations.APIService.RaiseExpectations(r.kvKey, 1, 0)
+		_, err := r.aggregatorclient.Create(context.Background(), apiService, metav1.CreateOptions{})
+		if err != nil {
+			r.expectations.APIService.LowerExpectations(r.kvKey, 1, 0)
+			return fmt.Errorf("unable to create apiservice %+v: %v", apiService, err)
 		}
 
-		injectOperatorMetadata(r.kv, &apiService.ObjectMeta, version, imageRegistry, id, true)
-		if !exists {
-			r.expectations.APIService.RaiseExpectations(r.kvKey, 1, 0)
-			_, err := r.aggregatorclient.Create(context.Background(), apiService, metav1.CreateOptions{})
-			if err != nil {
-				r.expectations.APIService.LowerExpectations(r.kvKey, 1, 0)
-				return fmt.Errorf("unable to create apiservice %+v: %v", apiService, err)
-			}
-		} else {
-			if !objectMatchesVersion(&cachedAPIService.ObjectMeta, version, imageRegistry, id, r.kv.GetGeneration()) || !certsMatch {
-				spec, err := json.Marshal(apiService.Spec)
-				if err != nil {
-					return err
-				}
+		return nil
+	}
 
-				ops, err := getPatchWithObjectMetaAndSpec([]string{}, &apiService.ObjectMeta, spec)
-				if err != nil {
-					return err
-				}
-				_, err = r.aggregatorclient.Patch(context.Background(), apiService.Name, types.JSONPatchType, generatePatchBytes(ops), metav1.PatchOptions{})
-				if err != nil {
-					return fmt.Errorf("unable to patch apiservice %+v: %v", apiService, err)
-				}
-				log.Log.V(2).Infof("apiservice %v updated", apiService.GetName())
+	modified := resourcemerge.BoolPtr(false)
+	resourcemerge.EnsureObjectMeta(modified, &cachedAPIService.ObjectMeta, apiService.ObjectMeta)
+	serviceSame := equality.Semantic.DeepEqual(cachedAPIService.Spec.Service, apiService.Spec.Service)
+	certsSame := equality.Semantic.DeepEqual(apiService.Spec.CABundle, cachedAPIService.Spec.CABundle)
+	prioritySame := cachedAPIService.Spec.VersionPriority == apiService.Spec.VersionPriority && cachedAPIService.Spec.GroupPriorityMinimum == apiService.Spec.GroupPriorityMinimum
+	insecureSame := cachedAPIService.Spec.InsecureSkipTLSVerify == apiService.Spec.InsecureSkipTLSVerify
+	// there was no change to metadata, the service and priorities were right
+	if !*modified && serviceSame && prioritySame && insecureSame && certsSame {
```

Comment:
does generation not work for apiServices? I'm curious why each field needs to be compared individually here.

Reply:
Correct, unfortunately there's no generation set for the APIService resource. We have to instead validate each field that we care about.

Reply:
I'm concerned the list of things we'll need to reconcile might change over time. Do we have a way to track what is and is not supposed to be managed?

Reply:
> Correct, unfortunately there's no generation set for the APIService resource. We have to instead validate each field that we care about.

Bummer. What you've done probably works all the time without issue. I'm just unsure how comfortable I am with the idea.

For example, if there's a mutation webhook that adds a label or annotation to the apiservice the operator posts, then the operator would be in a constant battle to remove it. With the generation id, i think we'd only be in battle if something calls Patch/Update after the fact. 

Maybe that webhook issue would never happen in practice. Knowing things like that are possible doesn't instill confidence that we're doing the right thing though.

Reply:
> For example, if there's a mutation webhook that adds a label or annotation to the apiservice the operator posts, then the operator would be in a constant battle to remove it. With the generation id, i think we'd only be in battle if something calls Patch/Update after the fact.

Labels and annotations won't be an issue unless they change something that we set/require in our definition template. But that is the case with all resources we are reconciling they all check those objects using openshifts `resourcemerge.EnsureObjectMeta(...)` function. 

> I'm concerned the list of things we'll need to reconcile might change over time. Do we have a way to track what is and is not supposed to be managed?

The only thing we have is the [API specification](https://docs.openshift.com/container-platform/3.11/rest_api/apiregistration_k8s_io/apiservice-apiregistration-k8s-io-v1.html#specification). However, I do feel pretty confident with this check since it is a replication of how openshift is checking for reconciliation and the APIService resource is owned by openshift.